tensorflow/core/lib/strings/str_util.cc

```diff
@@ -416,6 +417,119 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
   }
 }
 
+Status SplitUTF8(StringPiece text, const string& delim,
+                 std::vector<string>* result) {
+  // Bytes    Byte 1    Byte 2    Byte 3    Byte 4
+  //   1     0xxxxxxx
+  //   2     110xxxxx  10xxxxxx
+  //   3     1110xxxx  10xxxxxx  10xxxxxx
+  //   4     11110xxx  10xxxxxx  10xxxxxx  10xxxxxx
+  result->clear();
+  result->reserve(text.size());
+  size_t off = 0, len = 0;
+  bool advance = true;
+  for (size_t i = 0; i < text.size(); ++i) {
+    if (off == i) {
+      if ((text[i] & 0x80) == 0x00) {
+        string entry = text.substr(off, 1).ToString();
+        if (delim == "") {
+          // If delim is "" then always advance
+          result->emplace_back(entry);
+        } else if (delim != entry) {
```

Comment:
Could this be made a little more compact by separating out the length determination from the emplacement? Something like:

```c++
size_t char_start = 0;
size_t len = 0;
for (size_t i = 0; i < text.size(); ++i) {
  if (char_start == i) {
    // Determine byte length of next character.
    const char c = text[i];
    if (c & 0x80 == 0x00) {
     len = 1;
    } else if (c & 0xE0 == 0xC0) {
      len = 2;
    } else if (c & 0xF0 == 0xE0) {
      len = 3;
    } else if (c & 0XF8 == 0xF0) {
      len = 4;
   } else {
    result->clear();
    return errors::InvalidArgument("Invalid UTF-8 encoding at byte ", i);
   }
   // Validate all intermediate bytes are valid UTF-8
   if (TF_PREDICT_FALSE((len > 1 && (text[i] & 0xC0) != 0x80)) {
      /*  return error */
   }
   // Reached the end of a character
   if (char_start + len == i + 1) {
     const StringPiece entry = text.substr(char_start, len).ToString();
     char_start += len;
     if (delim == "") {
       result->emplace_back(entry);
     } else if (delim == entry) {
        advance = true;
     } else if (advance) {
        advance = false;
        result->emplace_back(entry);
    } else {
        result->back() = result->back() + entry;
   }
}
```

Reply:
Thanks. Done.

Reply:
Hmm...didn't see any change here. Perhaps a forgotten commit?

Reply:
Hmm...I didn't see much of a change. By switching things around a bit we won't have to treat the `len == 1` case differently from the others repeat the blocks of code that add to `result`. Something like this:

```c++
Status SplitUTF8(StringPiece text, const string& delim, std::vector<string>* result) {
  result->clear();
  size_t char_start = 0, char_len = 0;
  bool advance = true;
  for (size_t i = 0; i < text.size(); ++i) {
    if (char_start == i) {
      Status s = UTF8CharNumBytes(text, &char_len);
      if (!s.ok()) {
        result->clear();
        return errors::InvalidArgument("Invalid UTF8 encoding at byte ", i, ": ", s.error_message());
      }
    }
    // Validate all intermediate bytes are valid UTF-8
    if (TF_PREDICT_FALSE((len > 1 && (text[i] & 0xC0) != 0x80)) {
       return errors::InvalidArgument("Invalid UTF8 encoding at byte ", i);
    }
    // Reached the end of a character
    if (char_start + char_len == i + 1) {
      const StringPiece entry = text.substr(char_start, char_len).ToString();
      char_start += len;
      if (delim == "") {
        result->emplace_back(entry);
      } else if (delim == entry) {
         advance = true;
      } else if (advance) {
         advance = false;
         result->emplace_back(entry);
      } else {
         result->back() = result->back() + entry;
      }
    }
  }
  if (TF_PREDICT_FALSE(char_start < text.size())) {
    return errors::InvalidArgument("Invalid UTF8 encoding, incomplete last character");
  }
  return Status::OK();
}
```